tcpdevmem_cuda.cu

```diff
@@ -0,0 +1,452 @@
+  tmbuf->bytes_received = 0;
+  tmbuf->bytes_sent = 0;
+
+  cudaMalloc(&tmbuf->gpu_rx_mem_, opts->buffer_size);
+  cudaMalloc(&tmbuf->gpu_scatter_list_, opts->buffer_size);
+  tmbuf->rx_blks_ = new std::vector<TcpdRxBlock>();
+  tmbuf->scattered_data_ = new std::vector<long3>();
+  return 0;
+}
+
+int tcpd_send(int socket, void *buf, size_t n, int flags, struct thread *t) {
+  int gpu_mem_fd_;
+  struct iovec iov;
+  struct msghdr msg;
+  struct cmsghdr *cmsg;
+  char offsetbuf[CMSG_SPACE(sizeof(uint32_t) * 2)];
+  struct tcpdevmem_cuda_mbuf *tmbuf;
+
+  if (!buf) return -1;
+
+  tmbuf = (struct tcpdevmem_cuda_mbuf *)buf;
+  gpu_mem_fd_ = tmbuf->gpu_mem_fd_;
+
+  memset(&msg, 0, sizeof(msg));
+
+  iov.iov_base = NULL;
+  iov.iov_len = n - tmbuf->bytes_sent;
+
+  msg.msg_iov = &iov;
+  msg.msg_iovlen = 1;
+
+  msg.msg_control = offsetbuf;
+  msg.msg_controllen = sizeof(offsetbuf);
+
+  cmsg = CMSG_FIRSTHDR(&msg);
+  cmsg->cmsg_level = SOL_SOCKET;
+  cmsg->cmsg_type = SCM_DEVMEM_OFFSET;
+  cmsg->cmsg_len = CMSG_LEN(sizeof(int) * 2);
+  *((int*)CMSG_DATA(cmsg)) = gpu_mem_fd_;
+  ((int *)CMSG_DATA(cmsg))[1] = (int)tmbuf->bytes_sent;
+
+  ssize_t bytes_sent = sendmsg(socket, &msg, MSG_ZEROCOPY | MSG_DONTWAIT);
+  if (bytes_sent < 0 && errno != EWOULDBLOCK && errno != EAGAIN)
+    PLOG_FATAL(t->cb, "sendmsg");
+
+  if (bytes_sent == 0)
+    PLOG_FATAL(t->cb, "sendmsg sent 0 bytes");
+
+  tmbuf->bytes_sent += bytes_sent;
+  if (tmbuf->bytes_sent == n)
+    tmbuf->bytes_sent = 0;
+
+  return bytes_sent;
+}
+
+int tcpd_recv(int socket, void *f_mbuf, size_t n, int flags, struct thread *t) {
+  struct iovec iov;
+  struct msghdr msg_local;
+  struct msghdr *msg;
+  struct tcpdevmem_cuda_mbuf *tmbuf;
+  int ret, client_fd;
+  int buffer_size = n;
+  size_t total_received = 0;
+  unsigned char *cpy_buffer;
+  const struct options *opts = t->opts;
+  std::vector<devmemvec> *vectors;
+  std::vector<devmemtoken> *tokens;
+  std::vector<TcpdRxBlock> *rx_blks_;
+  std::vector<long3> *scattered_data_;
+
+  if (!f_mbuf) return -1;
+
+  tmbuf = (struct tcpdevmem_cuda_mbuf *)f_mbuf;
+  cpy_buffer = (unsigned char *)tmbuf->cpy_buffer;
+  vectors = (std::vector<devmemvec> *)tmbuf->vectors;
+  tokens = (std::vector<devmemtoken> *)tmbuf->tokens;
+  rx_blks_ = (std::vector<TcpdRxBlock> *)tmbuf->rx_blks_;
+  scattered_data_ = (std::vector<long3> *)tmbuf->scattered_data_;
+
+  client_fd = socket;
+
+  char buf_dummy[n];
+  char offsetbuf[CMSG_SPACE(sizeof(iov) * 10000)];
+  msg = &msg_local;
+
+  memset(msg, 0, sizeof(struct msghdr));
+
+  iov.iov_base = buf_dummy;
+  iov.iov_len = n - tmbuf->bytes_received;
+  msg->msg_iov = &iov;
+  msg->msg_iovlen = 1;
+
+  msg->msg_control = offsetbuf;
+  msg->msg_controllen = sizeof(offsetbuf);
+
+  rx_blks_->clear();
+
+  ssize_t received = recvmsg(socket, msg, MSG_SOCK_DEVMEM | MSG_DONTWAIT);
+  if (received < 0 && (errno == EAGAIN || errno == EWOULDBLOCK)) {
+    printf("%s %d: recvmsg returned < 0\n", __func__, __LINE__);
```

Comment:
similar to elsewhere, I think we should follow the existing conventions in neper like using LOG_INFO even if we don't like it because it doesn't go to stdout, but no big deal for me.